### ISSUE TITLE
Closing dap-ui while repl is in insert mode breaks vim

### ISSUE THREAD

**Issue by jaghaimo** [rcarriga/nvim-dap-ui#208]
https://github.com/rcarriga/nvim-dap-ui/issues/188#issuecomment-1354291183 mentioned this is caused by setting `console` to `integratedTerminal`, which `nvim-dap-python` does.

Steps to reproduce:
- Debug any python file and toggle UI on (or just repl)
- Go to repl and enter insert mode
- Go back to your buffer while repl is in insert mode, everything works fine so far (if you go back to repl, it will still be in insert mode)
- While inside your buffer, toggle UI off
- Now every key you press will cause you to enter insert mode

**Comment by rcarriga** [comment]
Afraid I can't reproduce. Can you create a minimal init.lua?

Here's a starter one `nvim --clean -u min.lua`
```lua
-- ignore default config and plugins
vim.opt.runtimepath:remove(vim.fn.stdpath("config"))
vim.opt.packpath:remove(vim.fn.stdpath("data") .. "/site")
vim.opt.termguicolors = true
vim.opt.laststatus = 0

-- append test directory
local test_dir = "/tmp/nvim-config"
vim.opt.runtimepath:append(vim.fn.expand(test_dir))
vim.opt.packpath:append(vim.fn.expand(test_dir))

-- install packer
local install_path = test_dir .. "/pack/packer/start/packer.nvim"
local install_plugins = false

if vim.fn.empty(vim.fn.glob(install_path)) > 0 then
  vim.cmd("!git clone https://github.com/wbthomason/packer.nvim " .. install_path)
  vim.cmd("packadd packer.nvim")
  install_plugins = true
end

local packer = require("packer")

packer.init({
  package_root = test_dir .. "/pack",
  compile_path = test_dir .. "/plugin/packer_compiled.lua",
})

packer.startup(function(use)
  -- Packer can manage itself
  use("wbthomason/packer.nvim")

  use({ "mfussenegger/nvim-dap" })

  use({
    "rcarriga/nvim-dap-ui",
    requires = "mfussenegger/nvim-dap",
    after = "nvim-dap",
    config = function()
      require("dapui").setup({})
    end,
  })

  if install_plugins then
    packer.sync()
  end
end)

vim.cmd([[
command! DapUIToggle lua require("dapui").toggle()
]])
```

**Comment by jaghaimo** [comment]
I tried adding `nvim-dap-python` to the minimal config, but couldn't make it working. The problem occurs for adapters that set `console` to `intergratedTerminal` (https://github.com/mfussenegger/nvim-dap/blob/master/doc/dap.txt#L526) and won't happen without starting a debug session first. Afterwards, no active debug session is needed to trigger it.

**Comment by siraphobk** [comment]
I'm having the same issue. I believe it's the conflict between the plugins that causes this problem.

To work around on this, I force `dapui.setup()` every time I close the debug session.

```lua
-- dap should be setup before executing this function
local dap = require("dap")
local dapui = require("dapui")

local config = {
  layouts = {
    {
      elements = {
        { id = "scopes", size = 0.25 },
        "breakpoints",
        "stacks",
        "watches",
      },
      size = 40,
      position = "left",
    },
    {
      elements = { "repl" },
      size = 0.25, -- 25% of total lines
      position = "bottom",
    },
  },
}

dapui.setup(config)

dap.listeners.after.event_initialized["dapui_config"] = function()
  dapui.open({
    reset = true,
  })
end

dap.listeners.before.event_terminated["dapui_config"] = function()
  dapui.close({})
  dapui.setup(config)
end

dap.listeners.before.event_exited["dapui_config"] = function()
  dapui.close({})
  dapui.setup(config)
end
```

---

My investigation

I use 'kevinhwang91/nvim-ufo' which requires 'kevinhwang91/promise-async'. By disabling both plugins, the problem seems to disappear. So I believe these plugins might cause the conflict.

**Comment by rcarriga** [comment]
Interesting that it seems to be a conflict between plugins. nvim-dap-ui an neotest both use an async framework that should be compatible (but also should never need to interact) whereas promise-async is a completely different style. 

I am using all of these plugins without issue so unfortunately I can't debug further. @siraphobk are you able to reproduce using a minimal init.lua?

**Comment by siraphobk** [comment]
@rcarriga I might be wrong about those plugin conflicts tho. However, I'll try to setup using a minimal init.lua and will post it here.

**Comment by rcarriga** [comment]
I believe this was due to a neovim core bug, and has been fixed in nightly (possibly 9.1). Closing this but let me know if it's still an issue
